### ISSUE TITLE
Training hangs when using nccl in two pod on two nodes

### ISSUE THREAD

**Issue by boxwh1** [NVIDIA/nccl#193]
Just as the same as the issues we submit in https://github.com/pytorch/pytorch/issues/17703.
Distributed training hangs when using nccl with DistributedDataParallel mode in Pytorch in two pod on two nodes.
The problem occasionally happens and can not be reproduced each time.
The pods were started by k8s using flannel network. The physical nodes are connected by 10Ge ethernet.

Training started normally, but hanged in the middle of training process.
![image](https://user-images.githubusercontent.com/41627739/54351505-96be5d00-468a-11e9-9c03-422951fa279b.png)

Then there was no output again. The training process did not exit for servel hours.
We found that the GPU utilization was always 100% on both two nodes.
![image](https://user-images.githubusercontent.com/41627739/54351523-9e7e0180-468a-11e9-9d22-0aed21dd35cd.png)

After we added NCCL_DEBUG=WARN before the training command, we found the WARN as follows.
On on pod:
```
worker0-0d9gnb:15:91 [3] include/socket.h 398 NCCL WARN Call to write failed : Connection reset by peer.
worker0-0d9gnb:15:91 [3] transport.cu:153 NCCL WARN transport.cu:153 ->2 [Proxy thread error]
```
On the other pod:
```
worker0-0d9gnb:15:91 [3] include/socket.h 398 NCCL WARN Call to recv failed : Connection reset by peer.
worker0-0d9gnb:15:91 [3] transport.cu:153 NCCL WARN transport.cu:153 ->2 [Proxy thread error]
```

The problem blocks our model training for about ten days.
Would you please give us some advices?

Environment
Training code is pull from https://github.com/pytorch/examples/tree/master/imagenet.
Model is Resnet. We used 8 V100s to train (4 on each node).
Batch size was set 8, 32 or 128 per GPU.
The network rate is about 500 MB/s to synchronize the gradients.

PyTorch Version (e.g., 1.0): both 1.0 and 0.4.1
OS (e.g., Linux): Ubunut 16.04
How you installed PyTorch (conda, pip, source): pip
Build command you used (if compiling from source): None
Python version: 3.6
CUDA/cuDNN version: CUDA 9.0/cuDNN 7.1.2
GPU models and configuration: V100 with NVLink
Nccl: 2.3.7 (built-in Pytorch 0.4.1)

**Comment by sjeaugey** [comment]
This is the expected behavior when a networking error occurs (in your case : the socket being reset by peer).

NCCL 2.4 introduces two new calls : ncclCommGetAsyncError() and ncclCommAbort(). PyTorch should call ncclCommGetAsyncError() to catch the network error and then call ncclCommAbort().

Also starting from NCCL 2.4, and if the socket disconnect is a transient error, maybe we could try to immediately reconnect the socket in https://github.com/NVIDIA/nccl/blob/master/src/transport/net_socket.cu and continue.

But the simplest solution by far is to fix the network so that it no longer closes sockets.

**Comment by boxwh1** [comment]
Thanks for a quick reply.
We have tried to launch the pod to check the network connection.
We found the network was healthy when the training process was hanging.
It is likely that there is transient package loss between the network connection, but NCCL communication hangs and can not be recovered.

More interesting, we have done some experiment to check the influence of network for the distributed training.
First, we closed the network connection when the training is running normally, then the training hangs. But after we recover the network , the training process can be recovered, too. Does it mean that NCCL recover the socket?

However, in our issue, the training process hangs and can not be recovered when encountering the transient network package loss, even the network recovered quickly.

We will check the network in k8s and ensure the network will not close sockets as far as possibly.
But sometimes we may not be able to control it, especially when the network load is high.

**Comment by sjeaugey** [comment]
Whether the network comes back quickly or not doesn't really matter. If the socket was closed, the network communication fails and that is a fatal error for NCCL currently.

Now you could catch that error and abort instead of hanging but NCCL does not currently have the ability to recover once the socket has been closed.

**Comment by boxwh1** [comment]
Got it. Thanks a lot.

**Comment by sjeaugey** [comment]
Closing ; please reopen if you have further questions.

**Comment by 372046933** [comment]
> This is the expected behavior when a networking error occurs (in your case : the socket being reset by peer).
> 
> NCCL 2.4 introduces two new calls : ncclCommGetAsyncError() and ncclCommAbort(). PyTorch should call ncclCommGetAsyncError() to catch the network error and then call ncclCommAbort().
> 
> Also starting from NCCL 2.4, and if the socket disconnect is a transient error, maybe we could try to immediately reconnect the socket in https://github.com/NVIDIA/nccl/blob/master/src/transport/net_socket.cu and continue.
> 
> But the simplest solution by far is to fix the network so that it no longer closes sockets.

Link updated:
https://github.com/NVIDIA/nccl/blob/master/src/transport/net_socket.cc

**Comment by 372046933** [comment]
@sjeaugey On TF 2.0 and PyTorch, distributed training across multiple nodes is almost impossible. If network packet loss happended, what's the expected behavior of NCCL

**Comment by sjeaugey** [comment]
Network protocols (TCP and IB Verbs)  are supposed to operate reliably even if packets are dropped.

Are you experiencing issues with sockets closing, causing NCCL to fail ?

**Comment by 372046933** [comment]
@sjeaugey  I use `ss -an` to check the connection information. Only find a single direction connection. For example, I have node `A, B, C`. On `A`, `ss` shows connection to `B` is established, on 'B', connection to 'C' is established, and on 'C', there is connection to 'A'. But there is no connection in reverse direction. E.g. on 'A', no connection to `C` is shown

**Comment by sjeaugey** [comment]
I would suggest opening a new issue. If any connection fails, NCCL would return an error (through normal NCCL calls or through ncclGetAsyncError) and it would print an explicit message if you set NCCL_DEBUG=WARN.

**Comment by MrAta** [comment]
> but NCCL does not currently have the ability to recover once the socket has been closed.

@sjeaugey Does NCCL now have the ability to recover once the socket has been closed? if so, can you please point me to the commit(s) for that?

**Comment by sjeaugey** [comment]
No, it doesn't if a socket closes NCCL assumes this is a permanent issue and doesn't try to re-open it.
